lib/brook/serializer.ex

```diff
@@ -0,0 +1,49 @@
+defprotocol Brook.Event.Kafka.Serializer do
+  @type type :: atom()
+  @type reason :: term()
+  @fallback_to_any true
+
+  @spec serialize(term()) :: {:ok, type(), term()} | {:ok, term()} | {:error, reason()}
+  def serialize(data)
+end
+
+defprotocol Brook.Event.Kafka.Deserializer do
+  @type reason :: term()
+  @fallback_to_any true
+
+  @spec deserialize(module(), term()) :: {:ok, term()} | {:error, reason()}
+  def deserialize(struct, data)
+end
+
+defimpl Brook.Event.Kafka.Serializer, for: Any do
+  def serialize(%custom_struct{} = data) do
+    case data |> Map.from_struct() |> serialize() do
+      {:ok, serialized_data} -> {:ok, custom_struct, serialized_data}
+      result -> result
+    end
+  end
+
+  def serialize(data) do
+    Jason.encode(data)
+  end
+end
+
+defimpl Brook.Event.Kafka.Deserializer, for: Any do
+  def deserialize(:undefined, data) do
```

Comment:
At what point are you passing around the `:undefined` atom? Is that too arbitrary if you’re expecting the client to implement something?

Reply:
:undefined is an internal atom to indicate no struct was defined for this message, would something like :no_struct be better?

Reply:
Didn't see where the other instance was being thrown. Since this is all internal, I don't think it matters and personally would prefer `:undefined`